### PR TITLE
Test search page, creating tests

### PR DESCRIPTION
**Pull request by grunwjir**
Test search page was improved, it supports list and card view. UI changes. 

The page for test detail was created, it is possible to add/edit/remove metrics and alerts. Subscribing to test alerts is working. UI of this page is temporary, will be changed.

Form validation for tests, metrics and alerts was implemented.

### CODE REVIEW

File: model/src/main/java/org/perfrepo/model/to/OrderBy.java [hunk 1]
```diff
@@ -27,8 +27,8 @@
 
    UID_DESC,
 
-   GROUP_ID_ASC,
+   GROUP_ASC,
 
-   GROUP_ID_DESC;
+   GROUP_DESC;
```
Comment (grunwjir) [inline model/src/main/java/org/perfrepo/model/to/OrderBy.java:33]: Can I rename this?

Comment (Holmistr) [reply model/src/main/java/org/perfrepo/model/to/OrderBy.java:33]: It's actually better if you rename this :)

File: model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionGroupValueDto.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package org.perfrepo.dto.test_execution;
+
+import java.util.List;
+
+/**
+ * @author Jiri Grunwald (grunwjir@gmail.com)
+ */
+public class TestExecutionGroupValueDto {
```
Comment (Holmistr) [inline model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionGroupValueDto.java]: Could you please explain purpose of this class? I would also rename it to ValueGroupDto, you don't have to put "TestExecution" everywhere

File: model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionSearchParams.java [hunk 3]
```diff
@@ -1,10 +1,73 @@
...
+import org.perfrepo.model.to.OrderBy;
+
+import java.util.Set;
+
 /**
  * Data transfer object that contains filter parameters for {@link TestExecutionDto} searching query.
  *
  * @author Jiri Grunwald (grunwjir@gmail.com)
  */
 public class TestExecutionSearchParams {
```
Comment (Holmistr) [inline model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionSearchParams.java]: I think TestExecutionSearchCriteria is more descriptible, but that's a minor comment.

File: model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionValueDto.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+package org.perfrepo.dto.test_execution;
+
+import java.util.Map;
+
+/**
+ * @author Jiri Grunwald (grunwjir@gmail.com)
+ */
+public class TestExecutionValueDto {
```
Comment (Holmistr) [inline model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionValueDto.java]: Rename to ValueDto. Values are present only for test execution, it's obvious.

File: web/pom.xml [hunk 5]
```diff
@@ -306,26 +306,48 @@
          </plugin>
 
          <plugin>
-            <groupId>org.codehaus.mojo</groupId>
-            <artifactId>exec-maven-plugin</artifactId>
-            <version>1.3.2</version>
+            <groupId>com.github.eirslett</groupId>
+            <artifactId>frontend-maven-plugin</artifactId>
```
Comment (Holmistr) [inline web/pom.xml:310]: Could you please very shortly explain, what the plugin basically does? It's just I have never used it before, nothing wrong with the usage at all :)

File: web/src/main/java/org/perfrepo/web/adapter/dummy_impl/builders/TestExecutionValueDtoBuilder.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
+package org.perfrepo.web.adapter.dummy_impl.builders;
+
+import org.perfrepo.dto.test_execution.TestExecutionValueDto;
+
+import java.util.HashMap;
+
+/**
+ * @author Jiri Grunwald (grunwjir@gmail.com)
+ */
+public class TestExecutionValueDtoBuilder {
```
Comment (Holmistr) [inline web/src/main/java/org/perfrepo/web/adapter/dummy_impl/builders/TestExecutionValueDtoBuilder.java]: Rename, see previous comments.

File: web/src/main/java/org/perfrepo/web/rest/endpoints/TestExecutionRestApi.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+      return Response
+              .status(Response.Status.OK)
+              .header("X-Pagination-Total-Count", result.getTotalCount())
+              .header("X-Pagination-Current-Page", result.getCurrentPage())
+              .header("X-Pagination-Page-Count", result.getPageCount())
+              .header("X-Pagination-Per-Page", result.getPerPage())
+              .entity(result.getData()).build();
+   }
+
+
```
Comment (Holmistr) [inline web/src/main/java/org/perfrepo/web/rest/endpoints/TestExecutionRestApi.java]: Extra empty line

**Review by Holmistr — CHANGES_REQUESTED** [review]

**Comment by Holmistr** [comment]
@grunwjir I've placed some minor comments, but let me add some more high-level regarding tests pages:
* the listing is not expanding to full width on my monitor, looks like the width is fixed somehow and it looks IMHO weird (see attached screenshot)
* add headings to the columns of the table (like test name, test UID)
* the test name should be clickable on its own redirecting to the test detail
* There is button "Test detail". First - it can be just "Detail" with maybe different icon, but I would also create buttons "Edit" and "Delete" (also remove "test", it's obvious) and now hide them into dropdown menu. However, I really like the dropdown menu and the "Create test execution" and "Show test executions" should remain there, that's absolutely brilliant :)
* there is a small arrow on each row showing more information. To me it doesn't make sense to duplicate the information about the test there. However, I really like the idea of this dropdown and I think we can use it much better. It could show last x (let's say 3) test executions that were run for this test. Just a mini-table with 3 rows saying something like "Last 3 test executions" and listing - execution name, execution date and tags. That would be truly awesome, hopefully it's not too much work. And it's important that this information would be lazily-loaded, not preloaded for every test in advance. 
* Also the arrow is not vertically centered with the test name, so it looks a bit weird.
* regarding the card view - I cannot imagine a usecase how it could be useful in this situation, hence I would remove it

Moving on to test detail page:
* the right align of "Uid", "Group" etc. is done on purpose? Looks non-traditional to me :-D I would prefer left align. Also please use UID everywhere :)
* listings of metric should contain also the description of the metric, it's actually quite common that you look at that to find out, what it actually means.
* listings of alerts - basically the same as above, I would show name of the alert and description, everything else can be hidden until "Edit" is clicked.
* even here (test detail) the page is narrower on my screen, perhaps something more general is not set up correctly (see attached screenshot)

That's all I've got. I know that some of the comments are really minor, please don't take it bad, I think it's an awesome work, just want to have it perfect :)

snimek z 2017-01-11 13-35-29
snimek z 2017-01-11 13-35-33

File: model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionGroupValueDto.java (hunk 2 above)
Comment (grunwjir) [reply model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionGroupValueDto.java]: It represents a group of test execution values. This group is for each test metric, therefore the values are grouped by a metric.

File: model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionValueDto.java (hunk 4 above)
Comment (grunwjir) [reply model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionValueDto.java]: Ok, renamed.

File: web/pom.xml (hunk 5 above)
Comment (grunwjir) [reply web/pom.xml:310]: It downloads and installs Node, NPM locally for this project. It runs Bower and executes Gulp build task.

File: web/src/main/java/org/perfrepo/web/adapter/dummy_impl/builders/TestExecutionValueDtoBuilder.java (hunk 6 above)
Comment (grunwjir) [reply web/src/main/java/org/perfrepo/web/adapter/dummy_impl/builders/TestExecutionValueDtoBuilder.java]: Renamed.

**Comment by grunwjir** [comment]
Answers to your comments:
> - the listing is not expanding to full width on my monitor, looks like the width is fixed somehow and 
it looks IMHO weird (see attached screenshot)

Repaired to full width of a screen. It was set to maximum width 1140px, but it was bad idea. 

> - add headings to the columns of the table (like test name, test UID)

It is [list view component](https://www.patternfly.org/pattern-library/content-views/list-view/#/_design) and it does not support headings. I know that my solution is not ideal, will be improved.

> - the test name should be clickable on its own redirecting to the test detail
> - There is button "Test detail". First - it can be just "Detail" with maybe different icon, but I would also create buttons "Edit" and "Delete" (also remove "test", it's obvious) and now hide them into dropdown menu. However, I really like the dropdown menu and the "Create test execution" and "Show test executions" should remain there, that's absolutely brilliant :)

Ok, resolved.

> - there is a small arrow on each row showing more information. To me it doesn't make sense to duplicate the information about the test there. However, I really like the idea of this dropdown and I think we can use it much better. It could show last x (let's say 3) test executions that were run for this test. Just a mini-table with 3 rows saying something like "Last 3 test executions" and listing - execution name, execution date and tags. That would be truly awesome, hopefully it's not too much work. And it's important that this information would be lazily-loaded, not preloaded for every test in advance.
>  - Also the arrow is not vertically centered with the test name, so it looks a bit weird.

I agree and show last x test executions is great idea, I will add it (next PR).

> - regarding the card view - I cannot imagine a usecase how it could be useful in this situation, hence I would remove it

The card view was removed.

> Moving on to test detail page:
...

Test detail page will be completely redesigned, next PR.

Thank you very much for your revision :)

File: model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionSearchParams.java (hunk 3 above)
Comment (grunwjir) [reply model/src/main/java/org/perfrepo/dto/test_execution/TestExecutionSearchParams.java]: Renamed

**Comment by Holmistr** [comment]
Changes looks good to me, I have noticed two more things:
* I would also want to see to which group the tests is assigned in the listings, hence add another column like "Group".
* I noticed that when I deploy the application to the server and access the "base" URL, the dashboard can be seen without login. However when I want to access anything else, login is finally required. This is something that we must not forget to fix :)

Please add at least the group column, note somewhere the login thing and we're good to integrate this PR.

**Pull request merged**